### PR TITLE
petsc: removed default unfree dependencies

### PR DESCRIPTION
**Pull request by yl3dy**
## Description of changes

PETSc itself is opensource, however ParMETIS is not, which breaks building by default with unfree disabled. Added an off-by-default flag to build with METIS/ParMETIS.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by yl3dy** [comment]
Result of `nixpkgs-review pr 336781` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages built:</summary>
  <ul>
    <li>getdp</li>
    <li>petsc</li>
    <li>precice</li>
    <li>python311Packages.pyprecice</li>
    <li>python311Packages.pyprecice.dist</li>
    <li>python312Packages.pyprecice</li>
    <li>python312Packages.pyprecice.dist</li>
  </ul>
</details>

File: pkgs/applications/science/math/getdp/default.nix [hunk 1]
```diff
@@ -1,14 +1,15 @@
 { lib, stdenv, fetchurl, cmake, gfortran, blas, lapack, mpi, petsc, python3 }:
 
-stdenv.mkDerivation rec {
+let
+  mpiSupport = petsc.passthru.mpiSupport;
+in stdenv.mkDerivation rec {
```
Comment (eclairevoyant) [inline pkgs/applications/science/math/getdp/default.nix:5]: Not sure I see the point of this change.

Comment (yl3dy) [reply pkgs/applications/science/math/getdp/default.nix:5]: `mpiSupport` was moved to `passthru` in `petsc`, so fixed the path. To not pollute the derivation attrset also made it a variable not an attribute.

Comment (ehmry) [reply pkgs/applications/science/math/getdp/default.nix:5]: Should `mpiSupport` be overrideable from `getdp`? If so, it should stay an attribute and be referred to using the `finalAttrs` pattern.

Comment (yl3dy) [reply pkgs/applications/science/math/getdp/default.nix:5]: No, `mpiSupport` should normally be tied to the respective flag in PETSc. Having PETSc with MPI (the default in nixpkgs) and `mpiSupport = false` in getDP leads to build error.

**Pull request merged**